src/lib/decorators.ts

```diff
@@ -41,24 +41,73 @@ export type Constructor<T> = {
  *
  */
 export const customElement = (tagName: string) =>
-    (clazz: Constructor<HTMLElement>) => {
-      window.customElements.define(tagName, clazz);
-      // Cast as any because TS doesn't recognize the return type as being a
-      // subtype of the decorated class when clazz is typed as
-      // `Constructor<HTMLElement>` for some reason. `Constructor<HTMLElement>`
-      // is helpful to make sure the decorator is applied to elements however.
-      return clazz as any;
+    (classOrDescriptor: Constructor<HTMLElement>|ClassDescriptor) => {
+      if (typeof classOrDescriptor === 'function') {
```

Comment:
Can we separate out the legacy decorators from the standard ones in separate functions?

It'd make the code easier to read, validate, and type check. Can still export only one version, like:

```
export const customElement = (tagName: string) =>
  (classOrDescriptor: Constructor<HTMLElement>|ClassDescriptor) => {
    if (typeof classOrDescriptor === 'function') return experimentalTsCustomElement(classOrDescriptor);
    return standardCustomElement(classOrDescriptor);
  );
```

Reply:
Done